### PR TITLE
[SubTask][THTTP] Make abstract TFrontendService for thrift http protocol support

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

This PR mainly focus on refactoring
1. Use Apache Thrift Naming style for thrift-based classes which is `TXyzAbc`, e.g. `TFrontendService`
2. Created `TFrontendService` and move the `TCLIService.Iface` APIs and other common fields in `TFrontendService`, later we will use it to implement the `KyuubiTHttpFrontendService` in the server module
3. introduce a new conf also for server FEs to be able to pick IP first

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @ulysses-you @yanghua

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -823,13 +823,21 @@ object KyuubiConf {
...
+      .doc("(deprecated) " +
+        "When true, engine register with hostname to zookeeper. When spark run on k8s" +
         " with cluster mode, set to false to ensure that server can connect to engine")
       .version("1.3.0")
       .booleanConf
-      .createWithDefault(true)
+      .createWithDefault(false)
+
+  val FRONTEND_CONNECTION_URL_USE_HOSTNAME: ConfigEntry[Boolean] =
+    buildConf("frontend.connection.url.use.hostname")
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:837]: I think it's `backend` ? the original idea of this config is used for that spark engine register its hostname or ipaddress to zk.

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:837]: it's Frontend

Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:837]: ah, I see. This PR refactor the related code and unify the connect url. But it is actually used by both server and engine. seems both `frontend` and `backend ` are not good.

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:837]: this conf will be propagated to other frontends and connection URL is part of base FE trait

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TBinaryFrontendService.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      info(s"Initializing $name on ${serverAddr.getHostName}:${serverSocket.getLocalPort} with" +
+        s" [$minThreads, $maxThreads] worker threads")
+    } catch {
+      case e: Throwable =>
+        error(e)
+        throw new KyuubiException(
+          s"Failed to initialize frontend service on $serverAddr:$portNum.",
+          e)
+    }
+    super.initialize(conf)
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/TBinaryFrontendService.scala:98]: move to line 91 ?

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala [hunk 3]
```diff
@@ -17,12 +17,10 @@
 
 package org.apache.kyuubi.service
 
-class NoopThriftBinaryFrontendService(override val serverable: Serverable)
-  extends ThriftBinaryFrontendService("NoopThriftBinaryFrontendService") {
+class NoopTBinaryFrontendService(override val serverable: Serverable)
+  extends TBinaryFrontendService("NoopThriftBinaryFrontend") {
```
Comment (yanghua) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala:21]: `NoopThriftBinaryFrontend ` -> `NoopTBinaryFrontendService `?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala [hunk 4]
```diff
@@ -18,143 +18,75 @@
...
-    super.initialize(conf)
-  }
+  import TFrontendService._
+  private val started = new AtomicBoolean(false)
+  private lazy val hadoopConf: Configuration = KyuubiHadoopUtils.newHadoopConf(conf)
+  private lazy val serverThread = new NamedThreadFactory(getName, false).newThread(this)
+  private lazy val serverHost = conf.get(FRONTEND_THRIFT_BINARY_BIND_HOST)
+
+  protected val portKey: ConfigEntry[Int]
+  protected lazy val portNum: Int = conf.get(portKey)
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala:54]: it's overkill, just `protected  def portNum : Int` and let child override it.

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala (hunk 3 above)
Comment (yaooqinn) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala:21]: It's better to use NoopThriftBinaryFrontend, the name is mainly for logging state change, like `Servcie [NoopThriftBinaryFrontend] is stop`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TBinaryFrontendService.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/TBinaryFrontendService.scala:98]: playing trick here as we always call `super.initialize` at the beginning or at the end

**Review by ulysses-you — APPROVED** [review]
LGTM if test pass

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala [hunk 5]
```diff
@@ -18,143 +18,73 @@
...
-import org.apache.thrift.protocol.{TBinaryProtocol, TProtocol}
-import org.apache.thrift.server.{ServerContext, TServer, TServerEventHandler, TThreadPoolServer}
-import org.apache.thrift.transport.{TServerSocket, TTransport}
+import org.apache.hive.service.rpc.thrift.{TCancelDelegationTokenReq, TCancelDelegationTokenResp, TCancelOperationReq, TCancelOperationResp, TCLIService, TCloseOperationReq, TCloseOperationResp, TCloseSessionReq, TCloseSessionResp, TExecuteStatementReq, TExecuteStatementResp, TFetchResultsReq, TFetchResultsResp, TGetCatalogsReq, TGetCatalogsResp, TGetColumnsReq, TGetColumnsResp, TGetCrossReferenceReq, TGetCrossReferenceResp, TGetDelegationTokenReq, TGetDelegationTokenResp, TGetFunctionsReq, TGetFunctionsResp, TGetInfoReq, TGetInfoResp, TGetInfoValue, TGetOperationStatusReq, TGetOperationStatusResp, TGetPrimaryKeysReq, TGetPrimaryKeysResp, TGetResultSetMetadataReq, TGetResultSetMetadataResp, TGetSchemasReq, TGetSchemasResp, TGetTablesReq, TGetTablesResp, TGetTableTypesReq, TGetTableTypesResp, TGetTypeInfoReq, TGetTypeInfoResp, TOpenSessionReq, TOpenSessionResp, TProtocolVersion, TRenewDelegationTokenReq, TRenewDelegationTokenResp, TStatus, TStatusCode}
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala:27]: `import org.apache.hive.service.rpc.thrift._`

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala:27]: I just use a tool to rfmt, don't blame me

**Review by pan3793 — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala (hunk 5 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/TFrontendService.scala:27]: I suggest we pay less attention to the imports area for code reviewing and just leave it to the tools for the current status of the community activities and status . also cc @yanghua @ulysses-you @hddong @turboFei. 

It's both saving reviewers and author's time

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala (hunk 3 above)
Comment (yanghua) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/service/NoopTBinaryFrontendService.scala:21]: ok

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**